modules/rest/src/main/java/org/apache/ignite/internal/rest/auth/DelegatingAuthenticationProvider.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.rest.configuration.AuthConfiguration;
+import org.apache.ignite.internal.rest.configuration.AuthProviderView;
+import org.apache.ignite.internal.rest.configuration.AuthView;
+import org.jetbrains.annotations.Nullable;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+
+/**
+ * Implementation of {@link AuthenticationProvider}. Creates a list of {@link Authenticator} according to provided {@link AuthConfiguration}
+ * and updates them on configuration changes. Delegates {@link AuthenticationRequest} to the list of {@link Authenticator}.
+ */
+public class DelegatingAuthenticationProvider implements AuthenticationProvider, ConfigurationListener<AuthView> {
+
+    private static final IgniteLogger LOG = Loggers.forClass(DelegatingAuthenticationProvider.class);
+
+    private final AtomicBoolean authEnabled = new AtomicBoolean(false);
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    private final Lock readLock = rwLock.readLock();
+    private final Lock writeLock = rwLock.writeLock();
+
+    private List<Authenticator> authenticators = new CopyOnWriteArrayList<>();
+
+    public DelegatingAuthenticationProvider() {}
+
+    @Override
+    public Publisher<AuthenticationResponse> authenticate(HttpRequest<?> httpRequest, AuthenticationRequest<?, ?> authenticationRequest) {
+        return Flux.create(emitter -> {
+            readLock.lock();
+            try {
+                if (authEnabled.get()) {
+                    Optional<AuthenticationResponse> successResponse = authenticators.stream()
+                            .map(it -> it.authenticate(authenticationRequest))
+                            .filter(AuthenticationResponse::isAuthenticated)
+                            .findFirst();
+                    if (successResponse.isPresent()) {
+                        emitter.next(successResponse.get());
+                        emitter.complete();
+                    } else {
+                        emitter.error(AuthenticationResponse.exception());
+                    }
+                } else {
+                    emitter.next(AuthenticationResponse.success("Unknown"));
+                    emitter.complete();
+                }
+            } finally {
+                readLock.unlock();
+            }
+        }, FluxSink.OverflowStrategy.ERROR);
+    }
+
+    @Override
+    public CompletableFuture<?> onUpdate(ConfigurationNotificationEvent<AuthView> ctx) {
+        return CompletableFuture.runAsync(() -> refreshProviders(ctx.newValue()));
+    }
+
+    public boolean authEnabled() {
+        return authEnabled.get();
+    }
+
+    private void refreshProviders(@Nullable AuthView view) {
+        writeLock.lock();
+        try {
+            if (view == null || !view.enabled()) {
+                authEnabled.set(false);
```

Comment:
Why do we need the `authEnabled` field? We can identify that authentication is disabled if the `authenticators` list is empty

Reply:
In case users want to disable authentication and save the current list of providers for the future.